File: src/jp/lionas/alexa/fundra/StartIntentHandler.java
package jp.lionas.alexa.fundra;

import com.amazon.ask.dispatcher.request.handler.HandlerInput;
import com.amazon.ask.dispatcher.request.handler.RequestHandler;
import com.amazon.ask.model.*;
import jp.lionas.alexa.fundra.mufg.FundUtil;
import jp.lionas.alexa.fundra.mufg.MufgFund;
import jp.lionas.alexa.fundra.mufg.model.CodeItem;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import static com.amazon.ask.request.Predicates.intentName;
import static com.amazon.ask.request.Predicates.sessionAttribute;

public class StartIntentHandler implements RequestHandler {

    @Override
    public boolean canHandle(HandlerInput handlerInput) {
        return handlerInput.matches(intentName(Const.START_INTENT_NAME)
                .and(sessionAttribute(Const.STATE_KEY, Const.STATE_START).negate())
                .and(sessionAttribute(Const.STATE_KEY, Const.STATE_END).negate())) ||
                handlerInput.matches(intentName("AMAZON.StartOverIntent"));
    }

    @Override
    public Optional<Response> handle(HandlerInput handlerInput) {

        Map<String, Object> sessionAttributes = handlerInput.getAttributesManager().getSessionAttributes();
        sessionAttributes.put(Const.STATE_KEY, Const.STATE_START);
        sessionAttributes.put(Const.INDEX_KEY, 0);

        String fundName = "";
        Map<String, Slot> slots = SlotUtil.getSlots(handlerInput);
        if(slots != null) {
            fundName = slots.get(Const.QUERY_FUND_NAME).getValue();
        }

        List<CodeItem> found = MufgFund.findFund(fundName);
        final String NO_DATA =
                (fundName != null && !fundName.isEmpty()) ? String.format(Const.NOT_FOUND, fundName) : Const.NO_SPOKEN;

        if(found == null || found.size() == 0) {
            // データが0件
            sessionAttributes.put(Const.STATE_KEY, Const.STATE_END);
            return handlerInput.getResponseBuilder()
                    .withSpeech(NO_DATA)
                    .withSimpleCard(Const.SKILL_NAME, NO_DATA)
                    .build();
        } else if(found.size() == 1) {
            // 1件だけマッチした場合
            return FundUtil.getLastFund(handlerInput, sessionAttributes, found.get(0));
        } else {
            // 複数マッチした場合
            return FundUtil.getRepeatedFunds(handlerInput, sessionAttributes, found);
        }

    }

}

File: src/jp/lionas/alexa/fundra/StartIntentHandler.kt
package jp.lionas.alexa.fundra

import com.amazon.ask.dispatcher.request.handler.HandlerInput
import com.amazon.ask.dispatcher.request.handler.RequestHandler
import com.amazon.ask.model.Response
import jp.lionas.alexa.fundra.mufg.FundUtil
import jp.lionas.alexa.fundra.mufg.MufgFund
import java.util.Optional

import com.amazon.ask.request.Predicates.intentName
import com.amazon.ask.request.Predicates.sessionAttribute

class StartIntentHandler : RequestHandler {

    override fun canHandle(handlerInput: HandlerInput): Boolean {
        return handlerInput.matches(intentName(Const.START_INTENT_NAME)
                .and(sessionAttribute(Const.STATE_KEY, Const.STATE_START).negate())
                .and(sessionAttribute(Const.STATE_KEY, Const.STATE_END).negate())) ||
                handlerInput.matches(intentName("AMAZON.StartOverIntent"))
    }

    override fun handle(handlerInput: HandlerInput): Optional<Response> {

        val sessionAttributes = handlerInput.attributesManager.sessionAttributes
        sessionAttributes[Const.STATE_KEY] = Const.STATE_START
        sessionAttributes[Const.INDEX_KEY] = 0

        var fundName = ""
        val slots = SlotUtil.getSlots(handlerInput)
        slots?.let {
            fundName = it[Const.QUERY_FUND_NAME]?.value.toString()
            sessionAttributes[Const.FUND_NAME] = fundName
        }

        val found = MufgFund.findFund(fundName)
        val noData = if (!fundName.isEmpty()) String.format(Const.NOT_FOUND, fundName) else Const.NO_SPOKEN

        return when {

            found.isEmpty() -> {
                // データが0件
                sessionAttributes[Const.STATE_KEY] = Const.STATE_END
                handlerInput.responseBuilder
                        .withSpeech(noData)
                        .withSimpleCard(Const.SKILL_NAME, noData)
                        .build()
            }

            found.size == 1 -> {
                // 1件だけマッチした場合
                FundUtil.getLastFund(handlerInput, sessionAttributes, found[0])
            }

            else -> {
                // 複数マッチした場合
                FundUtil.getRepeatedFunds(handlerInput, sessionAttributes, found)
            }

        }

    }

}
